# Before your start:
- Read the README.md file
- Comment as much as you can and use the resources in the README.md file
- Happy learning!

In [46]:
# Import your libraries:
from sklearn.datasets import load_diabetes
import pandas as pd

# Challenge 1 - Explore the Scikit-Learn Datasets

Before starting to work on our own datasets, let's first explore the datasets that are included in this Python library. These datasets have been cleaned and formatted for use in ML algorithms.

First, we will load the diabetes dataset. Do this in the cell below by importing the datasets and then loading the dataset  to the `diabetes` variable using the `load_diabetes()` function ([documentation](https://scikit-learn.org/stable/modules/generated/sklearn.datasets.load_diabetes.html)).

In [2]:
# Your code here:
diabetes = load_diabetes()

Let's explore this variable by looking at the different attributes (keys) of `diabetes`. Note that the `load_diabetes` function does not return dataframes. It returns you a Python dictionary.

In [3]:
# Your code here:
diabetes.keys()

dict_keys(['data', 'target', 'DESCR', 'feature_names', 'data_filename', 'target_filename'])

#### The next step is to read the description of the dataset. 

Print the description in the cell below using the `DESCR` attribute of the `diabetes` variable. Read the data description carefully to fully understand what each column represents.

*Hint: If your output is ill-formatted by displaying linebreaks as `\n`, it means you are not using the `print` function.*

In [4]:
# Your code here:
print(diabetes.DESCR)

.. _diabetes_dataset:

Diabetes dataset
----------------

Ten baseline variables, age, sex, body mass index, average blood
pressure, and six blood serum measurements were obtained for each of n =
442 diabetes patients, as well as the response of interest, a
quantitative measure of disease progression one year after baseline.

**Data Set Characteristics:**

  :Number of Instances: 442

  :Number of Attributes: First 10 columns are numeric predictive values

  :Target: Column 11 is a quantitative measure of disease progression one year after baseline

  :Attribute Information:
      - Age
      - Sex
      - Body mass index
      - Average blood pressure
      - S1
      - S2
      - S3
      - S4
      - S5
      - S6

Note: Each of these 10 feature variables have been mean centered and scaled by the standard deviation times `n_samples` (i.e. the sum of squares of each column totals 1).

Source URL:
https://www4.stat.ncsu.edu/~boos/var.select/diabetes.html

For more information see:
Bra

#### Based on the data description, answer the following questions:

1. How many attributes are there in the data? What do they mean?

1. What is the relation between `diabetes['data']` and `diabetes['target']`?

1. How many records are there in the data?

# Enter your answer here:

1. Hay 10 atributos. Son valores numéricos (escalados) sobre los pacientes: edad, sexo y otros parámetros obtenidos del análisis médico.


2. diabetes['data'] es el conjunto de datos que contiene las variabes independientes y predictoras, es decir, estos 10 atributos. diabetes['target'] es la variable dependiente a predecir.


3. Hay un total de 442 registros.

#### Now explore what are contained in the *data* portion as well as the *target* portion of `diabetes`. 

Scikit-learn typically takes in 2D numpy arrays as input (though pandas dataframes are also accepted). Inspect the shape of `data` and `target`. Confirm they are consistent with the data description.

In [5]:
# Your code here:
print(diabetes['data'].shape)
print(diabetes['target'].shape)

(442, 10)
(442,)


# Challenge 2 - Perform Supervised Learning on the Dataset

The data have already been split to predictor (*data*) and response (*target*) variables. Given this information, we'll apply what we have previously learned about linear regression and apply the algorithm to the diabetes dataset.

#### Let's briefly revisit the linear regression formula:

```
y = β0 + β1X1 + β2X2 + ... + βnXn + ϵ
```

...where:

- X1-Xn: data 
- β0: intercept 
- β1-βn: coefficients 
- ϵ: error (cannot explained by model)
- y: target

Also take a look at the `sklearn.linear_model.LinearRegression` [documentation](https://scikit-learn.org/stable/modules/generated/sklearn.linear_model.LinearRegression.html).

#### In the cell below, import the `linear_model` class from `sklearn`. 

In [6]:
# Your code here:
from sklearn import linear_model

#### Create a new instance of the linear regression model and assign the new instance to the variable `diabetes_model`.

In [7]:
# Your code here:
X = diabetes['data']
y = diabetes['target']

#### Next, let's split the training and test data.

Define `diabetes_data_train`, `diabetes_target_train`, `diabetes_data_test`, and `diabetes_target_test`. Use the last 20 records for the test data and the rest for the training data.

In [17]:
# Your code here:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=20, random_state=5)

Fit the training data and target to `diabetes_model`. Print the *intercept* and *coefficients* of the model.

In [26]:
# Your code here:
diabetes_model = linear_model.LinearRegression().fit(X_train, y_train)
print('Score:', diabetes_model.score(X_train, y_train))
print('Coefficients:', diabetes_model.coef_)
print('Intercept:', diabetes_model.intercept_)

Score: 0.5105600160074484
Coefficients: [ -20.656948   -244.26286088  511.31446526  328.41719115 -762.21704757
  465.59140368   98.85160328  188.02264042  732.99055662   55.43863775]
Intercept: 152.19708686651254


#### Inspecting the results

From the outputs you should have seen:

- The intercept is a float number.
- The coefficients are an array containing 10 float numbers.

This is the linear regression model fitted to your training dataset.

#### Using your fitted linear regression model, predict the *y* of `diabetes_data_test`.

In [22]:
# Your code here:
diabetes_model.predict(X_test)

array([195.89418829, 218.57710938, 126.14270545, 156.92226891,
       199.2300262 , 126.64078541,  98.3758214 , 233.20642803,
       172.99343075, 217.02034861,  54.35442932, 202.27630664,
       179.95720895, 144.90430046, 224.07267845, 113.25260395,
       122.97795172,  60.47163762,  79.03558513, 142.8065803 ])

#### Print your `diabetes_target_test` and compare with the prediction. 

In [24]:
# Your code here:
print(y_test)

[281. 332. 178. 155. 137.  65.  31. 275. 138. 173.  90. 221. 198.  88.
 268. 107.  83.  63.  65.  93.]


#### Is `diabetes_target_test` exactly the same as the model prediction? Explain.

# Your explanation here:
No es exactamente el mismo. De hecho, los resultados son bastante diferentes. Entiendo que es debido a que:
* El modelo usado puede no ser el mejor. El score es muy bajo: 0,51.
* Hay que eliminar al alguna variable para mejorar el modelo.
* Faltan más registros para que el modelo tenga más datos en el entrenamiento.

# Bonus Challenge 1 - Hypothesis Testing with `statsmodels`

After generating the linear regression model from the dataset, you probably wonder: then what? What is the statistical way to know if my model is reliable or not?

Good question. We'll discuss that using Scikit-Learn in Challenge 5. But for now, let's use a fool-proof way by using the ([Linear Regression class of StatsModels](https://www.statsmodels.org/dev/regression.html)) which can also conduct linear regression analysis plus much more such as calcuating the F-score of the linear model as well as the standard errors and t-scores for each coefficient. The F-score and t-scores will tell you whether you can trust your linear model.

To understand the statistical meaning of conducting hypothesis testing (e.g. F-test, t-test) for slopes, read [this webpage](https://onlinecourses.science.psu.edu/stat501/node/297/) at your leisure time. We'll give you a brief overview next.

* The F-test of your linear model is to verify whether at least one of your coefficients is significantly different from zero. Translating that into the *null hypothesis* and *alternative hypothesis*, that is:

    ```
    H0 : β1 = β2 = ... = β10 = 0
    HA : At least one βj ≠ 0 (for j = 1, 2, ..., 10)
    ```

* The t-tests on each coefficient is to check whether the confidence interval for the variable contains zero. If the confidence interval contains zero, it means the null hypothesis for that variable is not rejected. In other words, this particular vaiable is not contributing to your linear model and you can remove it from your formula.

Read the documentations of [StatsModels Linear Regression](https://www.statsmodels.org/dev/regression.html) as well as its [`OLS` class](https://www.statsmodels.org/dev/generated/statsmodels.regression.linear_model.OLS.html) which stands for *ordinary least squares*.

#### In the next cell, analyze `diabetes_data_train` and `diabetes_target_train` with the linear regression model of `statsmodels`. Print the fit summary.

Your output should look like:

![statsmodels regression](../statsmodels.png)

In [45]:
# Your code here:

import statsmodels.api as sm
mod = sm.OLS(X_train, y_train)

new_X_train = sm.add_constant(X_train) # adding a constant

model = sm.OLS(y_train, new_X_train).fit()
predictions = model.predict(new_X_train) 

print_model = model.summary()
print(print_model)

                            OLS Regression Results                            
Dep. Variable:                      y   R-squared:                       0.511
Model:                            OLS   Adj. R-squared:                  0.499
Method:                 Least Squares   F-statistic:                     42.87
Date:                Tue, 24 Sep 2019   Prob (F-statistic):           9.24e-58
Time:                        17:10:11   Log-Likelihood:                -2279.2
No. Observations:                 422   AIC:                             4580.
Df Residuals:                     411   BIC:                             4625.
Df Model:                          10                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const        152.1971      2.645     57.533      0.0

## Interpreting hypothesis testing results

Answer the following questions in the cell below:

1. What is the F-score of your linear model and is the null hypothesis rejected?

1. Does any of the t-tests of the coefficients produce a confidence interval containing zero? What are they?

1. How will you modify your linear reguression model according to the test results above?

# Your answers here:

1. F-score = 42.87. Prob(F-statistic) = 9.24e-58, que es <=0,05, por lo que rechazamos la hipótesis nula.


2. Sí. Los coeficientes x1, x5, x6, x7, x8, x10. Esto significa que la hipótesis nula para estas variables NO se rechaza.


3. Como no rechazamos la hipótesis nula para estas variables, esto significa que β1 = β5= β6 = β7 = β8 = β10 = 0. Por tanto, las variables x1, x5, x6, x7, x8, x10 no están contribuyendo a nuestro modelo lineal y podemos eliminarlos de la fórmula.

# Challenge 3 - Peform Supervised Learning on a Pandas Dataframe

Now that we have dealt with data that has been formatted for scikit-learn, let's look at data that we will need to format ourselves.

In the next cell, load the `auto-mpg.csv` file included in this folder and assign it to a variable called `auto`.

In [55]:
# Your code here:
auto = pd.read_csv("../auto-mpg.csv")

Look at the first 5 rows using the `head()` function:

In [56]:
# Your code here:
auto.head()

mpg  cylinders  displacement  horse_power  weight  acceleration  \
0  18.0          8         307.0        130.0    3504          12.0   
1  15.0          8         350.0        165.0    3693          11.5   
2  18.0          8         318.0        150.0    3436          11.0   
3  16.0          8         304.0        150.0    3433          12.0   
4  17.0          8         302.0        140.0    3449          10.5   

   model_year                       car_name  
0          70  \t"chevrolet chevelle malibu"  
1          70          \t"buick skylark 320"  
2          70         \t"plymouth satellite"  
3          70              \t"amc rebel sst"  
4          70                \t"ford torino"

Evaluate the data to ensure that all numeric columns are correctly detected as such by pandas. If a column is misclassified as object, coerce it to numeric.

In [57]:
# Your code here:
auto.dtypes

mpg             float64
cylinders         int64
displacement    float64
horse_power     float64
weight            int64
acceleration    float64
model_year        int64
car_name         object
dtype: object

In [63]:
auto['car_name'].nunique()

305

In [69]:
# Hay 305 modelos diferentes de coches. Apicamos encoding "bag of words" para darle un id numérico a cada uno de ellos.

unique_car_names = list(auto['car_name'].unique())
auto['car_id'] = auto['car_name'].apply(lambda car_name: unique_car_names.index(car_name))
auto['car_id'].value_counts()

31     6
131    5
34     5
17     5
30     4
      ..
201    1
202    1
203    1
204    1
152    1
Name: car_id, Length: 305, dtype: int64

In [72]:
auto.head(40)

mpg  cylinders  displacement  horse_power  weight  acceleration  \
0   18.0          8         307.0        130.0    3504          12.0   
1   15.0          8         350.0        165.0    3693          11.5   
2   18.0          8         318.0        150.0    3436          11.0   
3   16.0          8         304.0        150.0    3433          12.0   
4   17.0          8         302.0        140.0    3449          10.5   
5   15.0          8         429.0        198.0    4341          10.0   
6   14.0          8         454.0        220.0    4354           9.0   
7   14.0          8         440.0        215.0    4312           8.5   
8   14.0          8         455.0        225.0    4425          10.0   
9   15.0          8         390.0        190.0    3850           8.5   
10  15.0          8         383.0        170.0    3563          10.0   
11  14.0          8         340.0        160.0    3609           8.0   
12  15.0          8         400.0        150.0    3761           9.5   
13  14.0          8         455.0        225.0    3086          10.0   
14  24.0          4         113.0         95.0    2372          15.0   
15  22.0          6         198.0         95.0    2833          15.5   
16  18.0          6         199.0         97.0    2774          15.5   
17  21.0          6         200.0         85.0    2587          16.0   
18  27.0          4          97.0         88.0    2130          14.5   
19  26.0          4          97.0         46.0    1835          20.5   
20  25.0          4         110.0         87.0    2672          17.5   
21  24.0          4         107.0         90.0    2430          14.5   
22  25.0          4         104.0         95.0    2375          17.5   
23  26.0          4         121.0        113.0    2234          12.5   
24  21.0          6         199.0         90.0    2648          15.0   
25  10.0          8         360.0        215.0    4615          14.0   
26  10.0          8         307.0        200.0    4376          15.0   
27  11.0          8         318.0        210.0    4382          13.5   
28   9.0          8         304.0        193.0    4732          18.5   
29  27.0          4          97.0         88.0    2130          14.5   
30  28.0          4         140.0         90.0    2264          15.5   
31  25.0          4         113.0         95.0    2228          14.0   
32  25.0          4          98.0          NaN    2046          19.0   
33  19.0          6         232.0        100.0    2634          13.0   
34  16.0          6         225.0        105.0    3439          15.5   
35  17.0          6         250.0        100.0    3329          15.5   
36  19.0          6         250.0         88.0    3302          15.5   
37  18.0          6         232.0        100.0    3288          15.5   
38  14.0          8         350.0        165.0    4209          12.0   
39  14.0          8         400.0        175.0    4464          11.5   

    model_year                          car_name  car_id  
0           70     \t"chevrolet chevelle malibu"       0  
1           70             \t"buick skylark 320"       1  
2           70            \t"plymouth satellite"       2  
3           70                 \t"amc rebel sst"       3  
4           70                   \t"ford torino"       4  
5           70              \t"ford galaxie 500"       5  
6           70              \t"chevrolet impala"       6  
7           70             \t"plymouth fury iii"       7  
8           70              \t"pontiac catalina"       8  
9           70            \t"amc ambassador dpl"       9  
10          70           \t"dodge challenger se"      10  
11          70            \t"plymouth 'cuda 340"      11  
12          70         \t"chevrolet monte carlo"      12  
13          70       \t"buick estate wagon (sw)"      13  
14          70         \t"toyota corona mark ii"      14  
15          70               \t"plymouth duster"      15  
16          70                    \t"amc hornet" 

In [75]:
auto.drop(['car_name'], axis=1, inplace=True)
auto.head()

mpg  cylinders  displacement  horse_power  weight  acceleration  \
0  18.0          8         307.0        130.0    3504          12.0   
1  15.0          8         350.0        165.0    3693          11.5   
2  18.0          8         318.0        150.0    3436          11.0   
3  16.0          8         304.0        150.0    3433          12.0   
4  17.0          8         302.0        140.0    3449          10.5   

   model_year  car_id  
0          70       0  
1          70       1  
2          70       2  
3          70       3  
4          70       4

What is the newest model year and the oldest model year?

In [76]:
# Your code here:
print(auto['model_year'].max())
print(auto['model_year'].min())

82
70


Check the dataset for missing values and remove all rows containing at least one missing value.

In [81]:
# Your code here:
auto.isnull().sum()

mpg             0
cylinders       0
displacement    0
horse_power     6
weight          0
acceleration    0
model_year      0
car_id          0
dtype: int64

In [82]:
auto.shape

(398, 8)

In [85]:
auto.dropna(inplace=True) # how='any by default
auto.isnull().sum()

mpg             0
cylinders       0
displacement    0
horse_power     0
weight          0
acceleration    0
model_year      0
car_id          0
dtype: int64

In [86]:
# Comprobamos que se han eliminado 6 filas, ya que los 6 nulls correspondían a la misma columna.
auto.shape

(392, 8)

Find the frequency table for the `cylinders` column using the `value_counts()` function. How many possible values of cylinders are there?

In [87]:
# Your code here:
auto['cylinders'].value_counts() # Salen 5 tipos de cilindrada.

4    199
8    103
6     83
3      4
5      3
Name: cylinders, dtype: int64

We would like to generate a linear regression model that will predict mpg. To do this, first drop the `car_name` column since it does not contain any quantitative data. Next separate the dataframe to predictor and response variables. Separate those into test and training data with 80% of the data in the training set and the remainder in the test set. 

Assign the predictor and response training data to `X_train` and `y_train` respectively. Similarly, assign the predictor and response test data to `X_test` and `y_test`.

*Hint: To separate data for training and test, use the `train_test_split` method we used in previous labs.*

In [89]:
# Your code here:

# Ya eliminé antes la columna car_name.

X = auto.drop(['mpg'], axis=1) # data (input)
y = auto['mpg'] # target (output)

X_train, X_test, y_train, y_test = train_test_split(X, y, train_size=0.80, random_state=5)

Now we will processed and peform linear regression on this data to predict the mpg for each vehicle. 

#### In the next cell, create an instance of the linear regression model and call it `auto_model`. Fit `auto_model` with your training data.

In [92]:
# Your code here:
import sklearn
auto_model = sklearn.linear_model.LinearRegression().fit(X_train, y_train)
print('Score:', auto_model.score(X_train, y_train))
print('Coefficients:', auto_model.coef_)
print('Intercept:', auto_model.intercept_)

Score: 0.8132011378300722
Coefficients: [-0.55447664  0.01271198  0.00636163 -0.00692039  0.16634892  0.63749543
  0.00325624]
Intercept: -7.76193691634321


El Score es bastante bueno. A continuación voy a comparar las predicciones resultantes de pasar al modelo las X_test, con las y_test reales. Podremos ver que los valores no difieren mucho unos de otros.

In [96]:
auto_model.predict(X_test)

array([18.09608028, 25.96346356, 30.50779853, 30.43511973, 17.15169663,
       25.88169268, 26.6987153 , 31.61003133, 11.29902561, 18.73690279,
       11.5763088 , 26.72496617, 31.62313595, 16.01030491, 21.36962769,
       10.43183911, 26.47885737, 19.11651538, 29.61905723, 27.06687669,
       28.64282951, 19.18405843, 27.9901806 , 27.83683791, 21.80394961,
       30.09658275, 34.41360124, 32.70761152, 25.59881305, 33.75341081,
       26.03883457, 28.62453214, 30.33409092, 28.36812858, 12.98889067,
       21.07979203, 25.76933072,  5.81160312, 31.99755017, 27.6422033 ,
       31.83995855, 25.10963204, 21.24901749, 15.26815528, 22.34702699,
       22.24144884, 28.3976426 , 30.04910372, 11.74250149, 31.40240071,
       34.53662265, 20.2825553 , 25.7448652 , 13.86768333, 15.53109161,
       25.53684319, 32.05410305, 26.42690005, 31.27683791, 20.05288757,
       16.75237971, 14.90364742, 29.73932283, 25.67781611, 31.69340801,
       20.49993589, 23.88310984, 32.00926788, 14.36769117, 26.12

In [95]:
print(y_test)

215    13.0
184    25.0
302    34.5
304    37.3
265    17.5
       ... 
140    14.0
356    32.4
263    17.7
49     23.0
348    37.7
Name: mpg, Length: 79, dtype: float64


# Challenge 4 - Evaluate the Model

In addition to evaluating your model with F-test and t-test, you can also use the *Coefficient of Determination* (a.k.a. *r squared score*). This method does not simply tell *yes* or *no* about the model fit but instead indicates how much variation can be explained by the model. Based on the r squared score, you can decide whether to improve your model in order to obtain a better fit.

You can learn about the r squared score [here](). Its formula is:

![R Squared](../r-squared.png)

...where:

* yi is an actual data point.
* ŷi is the corresponding data point on the estimated regression line.

By adding the squares of the difference between all yi-ŷi pairs, we have a measure called SSE (*error sum of squares*) which is an application of the r squared score to indicate the extent to which the estimated regression model is different from the actual data. And we attribute that difference to the random error that is unavoidable in the real world. Obviously, we want the SSE value to be as small as possible.

#### In the next cell, compute the predicted *y* based on `X_train` and call it `y_pred`. Then calcualte the r squared score between `y_pred` and `y_train` which indicates how well the estimated regression model fits the training data.

*Hint: r squared score can be calculated using `sklearn.metrics.r2_score` ([documentation](https://scikit-learn.org/stable/modules/generated/sklearn.metrics.r2_score.html)).*

In [ ]:
# Your code here:


#### Our next step is to evaluate the model using the test data. 

We would like to ensure that our model is not overfitting the data. This means that our model was made to fit too closely to the training data by being overly complex. If a model is overfitted, it is not generalizable to data outside the training data. In that case, we need to reduce the complexity of the model by removing certain features (variables).

In the cell below, use the model to generate the predicted values for the test data and assign them to `y_test_pred`. Compute the r squared score of the predicted `y_test_pred` and the oberserved `y_test` data.

In [ ]:
# Your code here:


#### Explaining the results

The r squared scores of the training data and the test data are pretty close (0.8146 vs 0.7818). This means our model is not overfitted. However, there is still room to improve the model fit. Move on to the next challenge.

# Challenge 5 - Improve the Model Fit

While the most common way to improve the fit of a model is by using [regularization](https://datanice.github.io/machine-learning-101-what-is-regularization-interactive.html), there are other simpler ways to improve model fit. The first is to create a simpler model. The second is to increase the train sample size.

Let us start with the easier option and increase our train sample size to 90% of the data. Create a new test train split and name the new predictors and response variables `X_train09`, `X_test09`, `y_train09`, `y_test09`.

In [ ]:
# Your code here:


Initialize a new linear regression model. Name this model `auto_model09`. Fit the model to the new sample (training) data.

In [ ]:
# Your code here:


Compute the predicted values and r squared score for our new model and new sample data.

In [ ]:
# Your code here:


Compute the r squared score for the smaller test set. Is there an improvement in the test r squared?

In [ ]:
# Your code here:


# Bonus Challenge 2 - Backward Elimination 

The main way to produce a simpler linear regression model is to reduce the number of variables used in the model. In scikit-learn, we can do this by using recursive feature elimination. You can read more about RFE [here](https://scikit-learn.org/stable/modules/generated/sklearn.feature_selection.RFE.html).

In the next cell, we will import RFE

In [ ]:
from sklearn.feature_selection import RFE

Follow the documentation and initialize an RFE model using the `auto_model` linear regression model. Set `n_features_to_select=3`

In [ ]:
# Your code here:


Fit the model and print the ranking

In [ ]:
# Your code here:


Feature importance is ranked from most important (1) to least important (4). Generate a model with the three most important features. The features correspond to variable names. For example, feature 1 is `cylinders` and feature 2 is `displacement`.

Perform a test-train split on this reduced column data and call the split data `X_train_reduced`, `X_test_reduced`, `y_test_reduced`, `y_train_reduced`. Use an 80% split.

In [ ]:
# Your code here:


Generate a new model called `auto_model_reduced` and fit this model. Then proceed to compute the r squared score for the model. Did this cause an improvement in the r squared score?

In [ ]:
# Your code here: 


#### Conclusion

You may obtain the impression from this lab that without knowing statistical methods in depth, it is difficult to make major progress in machine learning. That is correct. If you are motivated to become a data scientist, statistics is the subject you must be proficient in and there is no shortcut. 

Completing these labs is not likely to make you a data scientist. But you will have a good sense about what are there in machine learning and what are good for you. In your future career, you can choose one of the three tracks:

* Data scientists who need to be proficient in statistical methods.

* Data engineers who need to be good at programming.

* Data integration specialists who are business or content experts but also understand data and programming. This cross-disciplinary track brings together data, technology, and business and will be in high demands in the next decade.